### PR TITLE
Add Kalman filters

### PR DESCRIPTION
**Pull request by mamikonyana**

### CODE REVIEW

**Comment by mamikonyana** [comment]
@pbailis could you review this?

File: contrib/src/main/java/macrobase/analysis/stats/KalmanFlattenedMatrixFilter.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package macrobase.analysis.stats;
+
+import macrobase.util.AlgebraUtils;
+import org.apache.commons.math3.linear.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class KalmanFlattenedMatrixFilter extends KalmanVectorFilter {
```
Comment (pbailis) [inline contrib/src/main/java/macrobase/analysis/stats/KalmanFlattenedMatrixFilter.java]: Comment on what this class does?

File: contrib/src/main/java/macrobase/analysis/stats/KalmanScalarFilter.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+import macrobase.util.AlgebraUtils;
+import org.apache.commons.math3.linear.ArrayRealVector;
+import org.apache.commons.math3.linear.RealMatrix;
+import org.apache.commons.math3.linear.RealVector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class KalmanScalarFilter extends KalmanVectorFilter {
```
Comment (pbailis) [inline contrib/src/main/java/macrobase/analysis/stats/KalmanScalarFilter.java]: Ditto comment?

File: contrib/src/main/java/macrobase/analysis/stats/KalmanVectorFilter.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+
+public class KalmanVectorFilter {
+    private static final Logger log = LoggerFactory.getLogger(KalmanVectorFilter.class);
+    protected double qScale;
+    private final RealMatrix H; // measurement operator
+    protected RealMatrix state;
+    protected RealMatrix cov;
+
+    // TODO: Deprecate
```
Comment (pbailis) [inline contrib/src/main/java/macrobase/analysis/stats/KalmanVectorFilter.java]: Deprecate?

File: contrib/src/main/java/macrobase/analysis/stats/KalmanVectorFilter.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        cov = MatrixUtils.createRealIdentityMatrix(2);
+    }
+
+    protected RealVector measure(RealMatrix state) {
+        return state.getRowVector(0);
+    }
+
+    public RealVector step(RealVector observation, int time) {
+        RealVector g = new ArrayRealVector(2);
+        g.setEntry(0, 0.5 * time * time);
```
Comment (pbailis) [inline contrib/src/main/java/macrobase/analysis/stats/KalmanVectorFilter.java]: intuition/comment as to why 0.5 is the right thing here?

**Comment by pbailis** [comment]
@mamikonyana minor comments, mostly to clarify what's going on in the code.

File: contrib/src/main/java/macrobase/analysis/stats/KalmanFlattenedMatrixFilter.java (hunk 1 above)
Comment (mamikonyana) [inline contrib/src/main/java/macrobase/analysis/stats/KalmanFlattenedMatrixFilter.java]: done

**Comment by mamikonyana** [comment]
@pbailis, I addressed comments, could you take another look?

File: contrib/src/main/java/macrobase/analysis/stats/KalmanFlattenedMatrixFilter.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import macrobase.util.AlgebraUtils;
+import org.apache.commons.math3.linear.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Wrapper around KalmanVectorFilter, allows passing matrices to the Kalman Filter and receive matrices instead of vectors.
+ */
+public class KalmanFlattenedMatrixFilter extends KalmanVectorFilter {
```
Comment (pbailis) [inline contrib/src/main/java/macrobase/analysis/stats/KalmanFlattenedMatrixFilter.java]: Just curious, why not make this part of Kalman Filter?

Should at the least be in a kalman filter package.

File: contrib/src/main/java/macrobase/analysis/stats/KalmanScalarFilter.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import org.apache.commons.math3.linear.ArrayRealVector;
+import org.apache.commons.math3.linear.RealVector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Wrapper around KalmanVectorFilter, allows passing scalars to the Kalman Filter and receive scalars instead of vectors.
+ */
+public class KalmanScalarFilter extends KalmanVectorFilter {
```
Comment (pbailis) [inline contrib/src/main/java/macrobase/analysis/stats/KalmanScalarFilter.java]: I still have a hard time understanding what this is doing from the comment.

File: contrib/src/main/java/macrobase/analysis/stats/KalmanFlattenedMatrixFilter.java (hunk 5 above)
Comment (mamikonyana) [inline contrib/src/main/java/macrobase/analysis/stats/KalmanFlattenedMatrixFilter.java]: I like this api better, the filter has an internal state and operates on matrices or vectors or scalars alone. So it is clear which type of filter it is from the class.

**Comment by mamikonyana** [comment]
@pbailis updated

p.s. I can't find a request review button, so it's still showing as "changes requested"

**Pull request merged**